src-self-hosted/main.zig

```diff
@@ -805,7 +806,7 @@ fn fmtPathFile(
     if (stat.kind == .Directory)
         return error.IsDir;
 
-    const source_code = source_file.readAllAlloc(fmt.gpa, stat.size, max_src_size) catch |err| switch (err) {
+    const source_code = source_file.readAllAlloc(fmt.gpa, max_src_size) catch |err| switch (err) {
```

Comment:
This patch is an improvement in the sense that it now supports non-"normal" files which report a stat size of 0, however it is also a performance regression because some code will now make a redundant stat() syscall when one has already been done. This line from `zig fmt` was the motivating use case for readAllAlloc to take the stat_size as a parameter.

Would you be in support of having readAllAlloc take a `size_hint: ?usize` parameter, and only doing a stat() call when that parameter is `null`?

Reply:
Or, keep `readAllAlloc` simple and make that an option only to `readAllAllocOptions`, which the zig fmt code here can take advantage of.

Reply:
I wouldn't expect the double `stat()` to be a performance problem as the FS metadata is cached, the second access should be almost for free and, even if that was not the case the time needed to fetch the info is negligible when you take into account the memory allocation + IO reading overhead.

Adding a `size_hint: ?usize` to `readAllAllocOptions` is fine by me if you really want to avoid that extra syscall, just let me know.

One more nit, `readAllAllocOptions` is misleading as it doesn't do what it says on the tin: the seek pointer is never reset to zero so it's more a `readToEndAllocOptions`.

Reply:
I agree you are probably right about perf in practice about the double stat call. However, here are my arguments for this:
 * Caching FS metadata is an OS implementation detail, which may not necessarily be true for all OS's (remember additionally that we have a "bring-your-own-OS-layer" concept), and this code is cross-platform.
 * It's not *that* much more complicated to have this extra API layer. We have this pattern in many places.
 * I get warm fuzzy feelings from looking at straces and seeing no redundant syscalls

The name change sounds good, go for it :+1: